walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,53 +15,90 @@
 package com.google.sps;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   // returns a list of time ranges appropriate for the request
   // time complexity is O(G * E) where:
```

Comment:
Is time complexity different now? Should it be max of (E*#required attendees, E*#optional attendees)?

Reply:
If i'm not mistaken, it will be max(E*#required attendees, E*#optional attendees) + #optional_events * #normal_events, considering the last loop. Each insertion takes at most O(n), since I'm inserting each optional interval to a list of taken intervals (number of taken intervals is equal to number of events in worst case. The worst case is when no pair of events intersects)

Reply:
Actually I can do it even faster, if I sort optional intervals firstly and then merge optionals with required in O(n)

Reply:
Yeah I overlooked the last loop. Feel free to make it faster :)

Reply:
I already made it :) Though it's not exactly as I explained above - for now I use TreeSet to store events, so it sorts automatically when new item is inserted. After I get the required answer, I insert optional events to the TreeSet and run the algo once again, which gives me slightly faster time complexity. Now I'm working on what I suggested above